include/dpp/isa_detection.hpp

```diff
@@ -0,0 +1,256 @@
+
+			_mm256_storeu_si256(up_sampled_ptr, up_sampled_added);
+		}
+	}
+
+#elif T_AVX
+
+	/**
+	 * @brief Collect sixteen elements from data_in, apply gain and increment, and store the result in data_out.
+	 *        This version uses AVX instructions.
+	 *
+	 * @param data_in Pointer to the input array of opus_int32 values.
+	 * @param data_out Pointer to the output array of opus_int16 values.
+	 * @param current_gain The gain to be applied to the elements.
+	 * @param increment The increment value to be added to each element.
+	 */
+	inline void collect_sixteen_elements(opus_int32* data_in, opus_int16* data_out, float current_gain, float increment) {
+		const __m128 current_gain128 = _mm_set1_ps(current_gain);
+		const __m128 increment128 = _mm_set1_ps(increment);
+
+		for (int x = 0; x < 4; ++x) {
+			auto original_values = _mm_load_si128(reinterpret_cast<const __m128i*>(data_in) + x);
+			auto original_float_values = _mm_cvtepi32_ps(original_values);
+			auto original_increment = _mm_mul_ps(_mm_set1_ps(increment),
+				_mm_set_ps(0.0f * static_cast<float>(x), 1.0f * static_cast<float>(x), 2.0f * static_cast<float>(x), 3.0f * static_cast<float>(x)));
+			auto original_gain = _mm_set1_ps(current_gain);
+			auto final_gain = _mm_add_ps(original_gain, original_increment);
+			__m128 currentSamplesNew128{ _mm_mul_ps(original_float_values, final_gain) };
+			auto selector_values{ _mm_cmp_ps(currentSamplesNew128, _mm_set1_ps(0.0f), _CMP_GE_OQ) };
+			auto max_values{ _mm_max_ps(currentSamplesNew128, _mm_set1_ps(static_cast<float>(std::numeric_limits<opus_int16>::min()))) };
+			auto min_values{ _mm_min_ps(currentSamplesNew128, _mm_set1_ps(static_cast<float>(std::numeric_limits<opus_int16>::max()))) };
+
+			__m128 final_values{ _mm_blendv_ps(max_values, min_values, selector_values) };
+
+			__m128i current_samples_newer{ _mm_cvtps_epi32(final_values) };
+			__m128i pack128{ _mm_packus_epi32(_mm_extracti_si64(current_samples_newer, 0, 64), _mm_extracti_si64(current_samples_newer, 63, 64)) };
+
+			_mm_storeu_si128(reinterpret_cast<__m128i*>(data_out) + x, pack128);
+		}
+	}
+
+	/**
+	 * @brief Combine 4 elements from up_sampled_vector with 4 elements from decoded_data and store the result in up_sampled_vector.
+	 *        This version uses AVX instructions.
+	 *
+	 * @param up_sampled_vector Pointer to the array of opus_int32 values.
+	 * @param decoded_data Pointer to the array of opus_int16 values.
+	 * @param x Index to select a specific set of elements to combine.
+	 */
+	inline void combine_samples(opus_int32* up_sampled_vector, const opus_int16* decoded_data, size_t x) {
+		for (size_t y = 0; y < 4; ++y) {
+			__m128i* up_sampled_ptr = reinterpret_cast<__m128i*>(up_sampled_vector + (x * 8 * y));
+			const __m128i* decoded_ptr = reinterpret_cast<const __m128i*>(decoded_data + (x * 8 * y));
+
+			__m128i current_up_sampled = _mm_loadu_si128(up_sampled_ptr);
```

Comment:
can/should we 'using' these __m128 etc to make them more readable?

Reply:
I'm pretty sure all those reinterpret_cast are UB as well? Doesn't this violate the strict aliasing rule of C++?

Reply:
dont know, personally i'd use a C cast here, for clarity to indicate "we arent doing anything special, and we just treat this data as raw bits"... because thats essentially what AVX is doing, packing the raw bytes of multiple values into a native 512 or 256 bit register.

Reply:
C casts are basically the same as reinterpret_cast in this case, that falls into UB territory as well
If the data was obtained as a different type, type aliasing needs to be done via memcpy or it is UB

Reply:
pretty sure you cant memcpy these, they are intrinsics and are in registers not ram... it works fine in discordcoreapi, so something acceptable is at play (if its ub, it seems to work everywhere just out of dumb luck)

Reply:
Yeah this is generally one of the UBs that work on most major hardwares but will break on anything unconventional